File: Cpp/18.cpp
/*

 Given an array S of n integers, are there elements a, b, c, and d in S such
 that a + b + c + d = target? Find all unique quadruplets in the array which
 gives the sum of target.

 Note: The solution set must not contain duplicate quadruplets.

 For example, given array S = [1, 0, -1, 0, -2, 2], and target = 0.

 A solution set is:
 [
 [-1,  0, 0, 1],
 [-2, -1, 1, 2],
 [-2,  0, 0, 2]
 ]

 */
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        int n = nums.size();
        if (n < 4) {
            return res;
        }
        sort(nums.begin(), nums.end());

        for (int i = 0; i + 3 < n; ++i) {
            if (i > 0 && nums[i - 1] == nums[i]) {
                continue;
            }

            if (nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3] > target) {
                break;
            }

            if (nums[i] + nums[n - 3] + nums[n - 2] + nums[n - 1] < target) {
                continue;
            }

            for (int j = i + 1; j + 2 < n; ++j) {
                if (j > (i + 1) && nums[j - 1] == nums[j]) {
                    continue;
                }

                if (nums[i] + nums[j] + nums[j + 1] + nums[j + 2] > target) {
                    break;
                }

                if (nums[i] + nums[j] + nums[n - 2] + nums[n - 1] < target) {
                    continue;
                }

                int start = j + 1, end = n - 1;
                while (start < end) {
                    if (nums[i] + nums[j] + nums[start] + nums[end] == target) {
                        res.push_back(
                            {nums[i], nums[j], nums[start], nums[end]});
                        start++;
                        while (start < end
                               && (nums[start - 1] == nums[start])) {
                            start++;
                        }
                        end--;
                        while (start < end && (nums[end + 1] == nums[end])) {
                            end--;
                        }
                    } else if (nums[i] + nums[j] + nums[start] + nums[end]
                               < target) {
                        start++;
                    } else {
                        end--;
                    }
                }
            }
        }

        return res;
    }
};

File: Python3/18.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        n = len(nums)
        if n<4:
            return []
        res = []
        nums = sorted(nums)
        for i in range(n-3):
            if i>0 and nums[i-1] == nums[i]:
                continue
            if nums[i]+nums[i+1] + nums[i+2] + nums[i+3] > target:
                continue
            if nums[i]+nums[n-3]+nums[n-2]+nums[n-1] < target:
                continue
            for j in range(i+1, n-2):
                if j-1>i and nums[j-1] == nums[j]:
                    continue
                if nums[i] + nums[j] + nums[j+1] + nums[j+2] > target:
                    continue
                if nums[i] + nums[j] + nums[n-2] + nums[n-1] < target:
                    continue
                start, end = j+1, n-1
                while start < end:
                    if nums[i]+nums[j] + nums[start] + nums[end] == target:
                        res.append([nums[i], nums[j], nums[start], nums[end]])
                        start +=1
                        end -=1
                    elif nums[i]+nums[j] + nums[start] + nums[end] < target:
                        start +=1
                    else:
                        end -=1
                    while start < end and start-1>j and nums[start] == nums[start-1]:
                        start +=1
                    while start < end and end+1<n and nums[end] == nums[end+1]:
                        end -=1
        return res
                    



